src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java

```diff
@@ -1104,6 +1151,92 @@ private void assertDefaultParameterMetaData(ParameterMetaData pmd, int expectedP
     }
   }
 
+  @Test
+  public void test12_InsertReturningTestData() throws SQLException {
+    assumeFalse(
+        "Emulator does not support DML with returning clause",
+        EmulatorSpannerHelper.isUsingEmulator());
+    try (Connection connection = createConnection(env, database)) {
+      connection.setAutoCommit(false);
+      try (PreparedStatement ps =
+          connection.prepareStatement(getSingersInsertReturningQuery(dialect.dialect))) {
+        assertDefaultParameterMetaData(ps.getParameterMetaData(), 5);
+        for (Singer singer : createSingers()) {
+          singer.setPreparedStatement(ps, getDialect());
+          assertInsertSingerParameterMetadata(ps.getParameterMetaData());
+          ps.addBatch();
+          // check that adding the current params to a batch will not reset the metadata
+          assertInsertSingerParameterMetadata(ps.getParameterMetaData());
+        }
+        int[] results = ps.executeBatch();
+        for (int res : results) {
+          assertEquals(1, res);
+        }
+      }
+      try (PreparedStatement ps =
+          connection.prepareStatement(getAlbumsInsertReturningQuery(dialect.dialect))) {
+        assertDefaultParameterMetaData(ps.getParameterMetaData(), 4);
+        for (Album album : createAlbums()) {
+          ps.setLong(1, album.singerId);
+          ps.setLong(2, album.albumId);
+          ps.setString(3, album.albumTitle);
+          ps.setLong(4, album.marketingBudget);
+          assertInsertAlbumParameterMetadata(ps.getParameterMetaData());
+          ResultSet rs = ps.executeQuery();
```

Comment:
nit, here and elsewhere: always use a `ResultSet` in a try-with-resources block (or otherwise explicitly close it when you are done with it)